File: Assets/scripts/ProjectileAsset/Projectile.cs
using ICSharpCode.NRefactory.Ast;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace ProjectileAsset
{
    public enum FlightTrajectory { Linear, Perabolic }

    public static class Projectile
    {
        public static Vector3 CalculateTrajectory(float time, Vector3 startPosition, Vector3 direction, float gravityMultiplier, float speed)
        {
            float angle = Vector3.Angle(new Vector3(direction.x, 0, direction.z), direction);
            angle = direction.y >= 0 ? angle : -angle;

            return new Vector3(
                    x: startPosition.x + direction.x * time * speed,
                    y: speed * time * Mathf.Sin(angle * Mathf.Deg2Rad) - (4.9f * gravityMultiplier) * Mathf.Pow(time, 2) + startPosition.y,
                    z: startPosition.z + direction.z * time * speed
                );
        }

        public static bool CheckCollision(Vector3 position, Vector3 nextPosition)
        {
            return Physics.Linecast(position, nextPosition);
        }

        //POTENTIAL CLEANUP NEEDED
        public static PenetrationResult[] GetPenetrations(Vector3[] points)
        {
            var results = new List<PenetrationResult>();
            for (int i = 0; i < points.Length - 1; i++)
            {
                var forward =
                    Physics.RaycastAll(
                    origin: points[0],
                    direction: points[1] - points[0],
                    maxDistance: Vector3.Distance(points[0], points[1])).OrderBy(x => x.distance);

                var backward =
                    Physics.RaycastAll(
                    origin: points[1],
                    direction: points[0] - points[1],
                    maxDistance: Vector3.Distance(points[0], points[1])).OrderBy(x => x.distance).ToArray();

                results.AddRange(forward.Select((entry, j) =>
                {
                    var exit = backward.ToList().ElementAtOrDefault(backward.Length - 1 - j);
                    var thickness = float.PositiveInfinity;
                    if (exit.point != null)
                        thickness = Vector3.Distance(entry.point, exit.point) * 1000;

                    return new PenetrationResult(
                        thickness: thickness,
                        collider: entry.collider,
                        entryPoint: entry.point,
                        exitPoint: exit.point
                    );
                }));
            }
            return results.ToArray();
        }
    }
}

File: ProjectileFunctions/Projectile.fs
namespace ProjectileAsset

open UnityEngine
open System

module Projectile =
    let CalculateTrajectory time (startPosition : Vector3) (startDirection : Vector3) gravityMultiplier speed = 
        let angle = 
            let positiveAngle = Vector3.Angle(Vector3(startDirection.x, 0.0f, startDirection.z), startDirection)
    
            match startDirection.y >= 0.0f with
            | true -> positiveAngle
            | false -> -positiveAngle 
    
        Vector3
            ( startPosition.x + startDirection.x * time * speed,
              speed * time * Mathf.Sin(angle * Mathf.Deg2Rad) - (4.9f * gravityMultiplier) * Mathf.Pow(time, 2.0f) + startPosition.y,
              startPosition.z + startDirection.z * time * speed )

    let GetPenetrations (points : Vector3[]) = 
        let validate (forward : RaycastHit[]) (backwards : RaycastHit[]) =
            match forward.Length = backwards.Length with
            |true -> backwards
            |false ->
                let nextBackwards =
                    Physics.RaycastAll(
                        points.[2],
                        points.[1] - points.[2],
                        Vector3.Distance(points.[1], points.[2]))
                    |> Array.sortBy(fun x -> x.distance)
                    
                match nextBackwards.Length > 0 && forward.[forward.Length - 1].collider = nextBackwards.[nextBackwards.Length - 1].collider with
                |true -> backwards
                            |> List.ofArray
                            |> List.append [nextBackwards.[nextBackwards.Length - 1]]
                            |> Array.ofList
                | false -> backwards

        let forward =
            Physics.RaycastAll(
                points.[0],
                points.[1] - points.[0],
                Vector3.Distance(points.[0], points.[1]))
            |> Array.sortBy(fun x -> x.distance)
    
        let backwards = 
            Physics.RaycastAll(
                points.[1],
                points.[0] - points.[1],
                Vector3.Distance(points.[0], points.[1]))
            |> Array.sortBy(fun x -> x.distance)
            |> validate forward

        let extractPenetrationResults (entry : RaycastHit) (exit : Option<RaycastHit>) =
            let thickness =
                match exit with
                |Some exit -> Vector3.Distance(entry.point, exit.point) * 1000.0f
                |None -> Single.PositiveInfinity

            { thickness = thickness
              entryHit = entry
              exitHit = exit |> Option.toNullable }
    
        forward
        |> Array.mapi (fun i x -> extractPenetrationResults x (backwards |> Array.tryItem (backwards.Length - i - 1)))


